src/specialize.rs

```diff
@@ -7,124 +9,124 @@ extern crate alloc;
 #[cfg(feature = "std")]
 extern crate std as alloc;
 
-#[cfg(feature = "specialize")]
-use crate::BuildHasherExt;
-#[cfg(feature = "specialize")]
 use alloc::string::String;
-#[cfg(feature = "specialize")]
 use alloc::vec::Vec;
 
-/// Provides a way to get an optimized hasher for a given data type.
-/// Rather than using a Hasher generically which can hash any value, this provides a way to get a specialized hash
-/// for a specific type. So this may be faster for primitive types.
-pub(crate) trait CallHasher {
-    fn get_hash<H: Hash + ?Sized, B: BuildHasher>(value: &H, build_hasher: &B) -> u64;
+#[inline]
+fn is<Generic, Expected>() -> bool
+where
+    Generic: ?Sized,
+    Expected: ?Sized + 'static,
+{
+    typeid::of::<Generic>() == TypeId::of::<Expected>()
 }
 
-#[cfg(not(feature = "specialize"))]
-impl<T> CallHasher for T
+#[inline]
+fn as_ahash_state<B: BuildHasher>(state: &B) -> Option<&RandomState> {
+    if is::<B, RandomState>() {
+        Some(unsafe { &*(state as *const B as *const RandomState) })
+    } else {
+        None
+    }
+}
+
+#[inline]
+fn hash_as_u64<B, T>(state: &B, value: &T) -> u64
 where
+    B: BuildHasher,
     T: Hash + ?Sized,
 {
-    #[inline]
-    fn get_hash<H: Hash + ?Sized, B: BuildHasher>(value: &H, build_hasher: &B) -> u64 {
-        let mut hasher = build_hasher.build_hasher();
+    if let Some(state) = as_ahash_state(state) {
```

Comment:
I don't think CallHasher actually needs to be generic with respect to BuildHasher, if we can change it to just take a RandomState we can eliminate these ifs

Reply:
Good call. I opened https://github.com/tkaitchuck/aHash/pull/261 to change `CallHasher`.